### PR TITLE
Enhance: Check time skew across nodes (hariharan-devarajan#160)

### LINKED ISSUE #160: Report warning when there is time skew across multiple nodes
Check if there is a time skew across multiple nodes and report a warning message in `analyzer.summary()`.

### PR DESCRIPTION
**Pull request by Jye-525**
This PR enhances the summary to check time skew across nodes (hariharan-devarajan Fixes #160 )

### CODE REVIEW

File: dfanalyzer/main.py [hunk 1]
```diff
@@ -542,9 +542,19 @@ def _remove_numbers(self, string_items):
...
+           logging.warn(f"The time skew exceeds {max_time_skew // 1e6} sec across hosts {hosts_ts_df.index.tolist()}")
+
     def summary(self):
         num_events = len(self.events)
         logging.info(f"Total number of events in the workload are {num_events}")
+        if len(self.events["hostname"].unique().compute()) > 1:
+            self._check_hosts_time_skew()
         total_time, total_io_time, total_compute_time, total_app_io_time, \
         only_io, only_compute, only_app_io, only_app_compute = self._calculate_time() #(0, 0, 0, 0, 0, 0, 0, 0, 0)
         hosts_used, filenames_accessed, num_procs, compute_tid, posix_tid, io_by_operations = dask.compute(
```
Comment (hariharan-devarajan) [inline dfanalyzer/main.py:560]:
```suggestion
        total_time, total_io_time, total_compute_time, total_app_io_time, \
        only_io, only_compute, only_app_io, only_app_compute = self._calculate_time() #(0, 0, 0, 0, 0, 0, 0, 0, 0)
        hosts_used, filenames_accessed, num_procs, compute_tid, posix_tid, io_by_operations = dask.compute(
            self.events["hostname"].unique(),
            self.events["filename"].unique(),
            self.events["pid"].unique(),
            self.events.query("phase == 1")["tid"].unique(),
            self.events.query("phase == 2")["tid"].unique(),
            self.events.query("phase == 2").groupby(["name"]).agg(
                {"dur": [sum, "count"], "size": [sum, "mean", median_fun, min, max, percentile(.25), percentile(.75)]})
        )
        hosts_used = hosts_used.to_list()
        if len(hosts_used) > 1:
            self._check_hosts_time_skew()
        
```

**Review by hariharan-devarajan — CHANGES_REQUESTED** [review]
We calculate the hosts below. We can reuse it.

File: dfanalyzer/main.py [hunk 2]
```diff
@@ -542,9 +542,19 @@ def _remove_numbers(self, string_items):
...
+        # filter the hosts if time skew exceeds 30 seconds
+        max_time_skew = 30e6
+        if np.std(hosts_ts_df['ts']) > max_time_skew:
+           logging.warn(f"The time skew exceeds {max_time_skew // 1e6} sec across hosts {hosts_ts_df.index.tolist()}")
+
     def summary(self):
         num_events = len(self.events)
         logging.info(f"Total number of events in the workload are {num_events}")
+        if len(self.events["hostname"].unique().compute()) > 1:
+            self._check_hosts_time_skew()
```
Comment (hariharan-devarajan) [inline dfanalyzer/main.py:557]: could u move this past 571 `hosts_used = hosts_used.to_list()`
```
       if len(hosts_used) > 1:
            self._check_hosts_time_skew()
```

Comment (Jye-525) [reply dfanalyzer/main.py]: I submitted the changes, please use [6e020f8](https://github.com/hariharan-devarajan/dftracer/pull/163/commits/6e020f8af369dc819d4965a7f224803aeed69d0f)

File: dfanalyzer/main.py [hunk 3]
```diff
@@ -542,9 +542,17 @@ def _remove_numbers(self, string_items):
...
+        hosts_ts_df = self.events.groupby('hostname').agg({'ts': 'min'}).compute()
+        # filter the hosts if time skew exceeds 30 seconds
+        max_time_skew = 30e6
+        if np.std(hosts_ts_df['ts']) > max_time_skew:
+           logging.warn(f"The time skew exceeds {max_time_skew // 1e6} sec across hosts {hosts_ts_df.index.tolist()}")
+
     def summary(self):
         num_events = len(self.events)
-        logging.info(f"Total number of events in the workload are {num_events}")
+        logging.info(f"Total number of events in the workload are {num_events}")
```
Comment (hariharan-devarajan) [inline dfanalyzer/main.py:555]:
```suggestion
        logging.info(f"Total number of events in the workload are {num_events}")
```

**Review by hariharan-devarajan — APPROVED** [review]
Looks good thanks.

**Pull request merged**